File: jems-server/src/main/kotlin/io/cloudflight/jems/server/project/service/model/ProjectCallSettings.kt
package io.cloudflight.jems.server.project.service.model

import io.cloudflight.jems.api.call.dto.CallType
import io.cloudflight.jems.server.call.service.model.ApplicationFormFieldConfiguration
import io.cloudflight.jems.server.call.service.model.CallCostOption
import io.cloudflight.jems.server.call.service.model.ProjectCallFlatRate
import io.cloudflight.jems.server.programme.service.costoption.model.ProgrammeLumpSum
import io.cloudflight.jems.server.programme.service.costoption.model.ProgrammeUnitCost
import io.cloudflight.jems.server.programme.service.stateaid.model.ProgrammeStateAid
import java.time.ZonedDateTime

data class ProjectCallSettings(
    val callId: Long,
    val callName: String,
    val callType: CallType,
    val startDate: ZonedDateTime,
    val endDate: ZonedDateTime,
    val endDateStep1: ZonedDateTime?,
    val lengthOfPeriod: Int,
    val isAdditionalFundAllowed: Boolean,
    val flatRates: Set<ProjectCallFlatRate>,
    val lumpSums: List<ProgrammeLumpSum>,
    val unitCosts: List<ProgrammeUnitCost>,
    val stateAids: List<ProgrammeStateAid>,
    var applicationFormFieldConfigurations: MutableSet<ApplicationFormFieldConfiguration>,
    val preSubmissionCheckPluginKey: String?,
    val firstStepPreSubmissionCheckPluginKey: String?,
    val costOption: CallCostOption,
) {
    fun isCallStep1Closed(): Boolean {
        return if (endDateStep1 == null) {
            false
        }
        else {
            ZonedDateTime.now().isBefore(startDate) || ZonedDateTime.now().isAfter(endDateStep1)
        }
    }

    fun isCallStep2Closed(): Boolean {
        return ZonedDateTime.now().isBefore(startDate) || ZonedDateTime.now().isAfter(endDate)
    }
}

File: jems-ui/src/app/project/model/projectCallSettings.ts
import {ProgrammeLumpSum} from './lump-sums/programmeLumpSum';
import {ProgrammeUnitCost} from './programmeUnitCost';
import {CallFlatRateSetting} from './call-flat-rate-setting';
import {ApplicationFormFieldConfigurationDTO} from '@cat/api';

export class ProjectCallSettings {
  callId: number;
  callName: string;
  startDate: Date;
  endDate: Date;
  endDateStep1: Date;
  lengthOfPeriod: number;
  flatRates: CallFlatRateSetting;
  lumpSums: ProgrammeLumpSum[];
  unitCosts: ProgrammeUnitCost[];
  multipleFundsAllowed: boolean;
  applicationFormFieldConfigurations: ApplicationFormFieldConfigurationDTO[];

  constructor(callId: number, callName: string, startDate: Date, endDate: Date, endDateStep1: Date, lengthOfPeriod: number, flatRates: CallFlatRateSetting, lumpSums: ProgrammeLumpSum[], unitCosts: ProgrammeUnitCost[], multipleFundsAllowed: boolean, applicationFormFieldConfigurations: ApplicationFormFieldConfigurationDTO[]) {
    this.callId = callId;
    this.callName = callName;
    this.startDate = startDate;
    this.endDate = endDate;
    this.endDateStep1 = endDateStep1;
    this.lengthOfPeriod = lengthOfPeriod;
    this.flatRates = flatRates;
    this.lumpSums = lumpSums;
    this.unitCosts = unitCosts;
    this.multipleFundsAllowed = multipleFundsAllowed;
    this.applicationFormFieldConfigurations = applicationFormFieldConfigurations;
  }
}
